Add "Transform message format" tutorial (#8)
ballerina-guides/integration-tutorials: transform-message-formats/README.md

# Transform message formats

## What you'll build

> Let's develop a service that accepts requests to make an appointment at a hospital, takes the reuqest in one format, tranform that into another format and forwward to a backend hospital service. This integrates a backend service and successfully transform a meesage into another format.

Reviewer: ```suggestion
Let's develop a service that accepts requests to make an appointment at a hospital, takes the request in one format, transforms that into another format and forwards to a backend hospital service. This integrates a backend service and successfully transform a message into another format.
```

Please make sure all typos are fixed and there are no grammar issues.

> To implement this use case, you will develop a REST service with single resource using Visual Studio Code with Ballerina Swan Lake extension. The resource will recieve the user request, transform it into acceptable format, sends to the hospital backend service and retrive the response with the correct reservation details.
> 

Reviewer 2: ```suggestion
To implement this use case, you will develop a REST service with single resource using Visual Studio Code with Ballerina Swan Lake extension. The resource will receive the user request, transform it into acceptable format, sends to the hospital backend service and retrieve the response with the correct reservation details.

```

The flow is as follows

1. Receive a request with a JSON payload in the following form.
```json
{
    "name": "John Doe",
    "dob": "1940-03-19",
    "ssn": "234-23-525",
    "address": "California",
    "phone": "8770586755",
    "email": "johndoe@gmail.com",
    "doctor": "thomas collins",
    "hospital_id": "grandoaks",
    "hospital": "grand oak community hospital",
    "card_no": "7844481124110331",
    "appointment_date": "2017-04-02"
}
```
2. Transform the data into the following form.
```json
{
    "patient": {
        "name": "John Doe",
        "dob": "1940-03-19",
        "ssn": "234-23-525",
        "address": "California",
        "phone": "8770586755",
        "email": "johndoe@gmail.com",
        "card_no": "7844481124110331"
    },
    "doctor": "thomas collins",
    "hospital": "grand oak community hospital",
    "appointment_date": "2017-04-02"
}
```

> 3. Extract the `hospital_id` and send the POST reuqest to the correct URL to make the reservation.

Reviewer 2: ```suggestion
3. Extract the `hospital_id` and send the POST request to the correct URL to make the reservation.
```

> 4. Retrive the response from the hospital backend in the following form

Reviewer: ```suggestion
4. Retreive the response from the hospital backend which will be in the following form
```

```json
{
    "appointmentNumber": 8,
    "doctor": {
        "name": "thomas collins",
        "hospital": "grand oak community hospital",
        "category": "surgery",
        "availability": "9.00 a.m - 11.00 a.m",
        "fee": 7000.0
    },
    "patient": {
        "name": "John Doe",
        "dob": "1940-03-19",
        "ssn": "234-23-525",
        "address": "California",
        "phone": "8770586755",
        "email": "johndoe@gmail.com"
    },
    "fee": 7000.0,
    "hospital": "grand oak community hospital",
    "confirmed": false,
    "appointmentDate": "2017-04-02"
}
```

### Concepts covered

- REST API
- HTTP client
- Data Mapper

## Let's get started!

### Step 1: Set up the workspace

Install [Ballerina Swan Lake](https://ballerina.io/downloads/) and the [Ballerina Swan Lake VSCode extension](https://marketplace.visualstudio.com/items?itemName=wso2.ballerina) on VSCode.

### Step 2: Develop the service

> // TODO: Write the development process

Reviewer: Do you plan on adding the data mapper here?

    isolated resource function post categories/[string category]/reserve(HealthcareReservation payload)
            returns ReservationResponse|http:NotFound|http:InternalServerError {
        HospitalReservation hospitalReservation = transform(payload);

        ReservationResponse|http:ClientError resp =
                    hospitalServiceEP->/[payload.hospital_id]/categories/[category]/reserve.post(hospitalReservation);

        if resp is ReservationResponse {
            log:printDebug("Reservation request successful",
                            name = payload.name,
                            appointmentNumber = resp.appointmentNumber);
            return resp;
        }

        log:printError("Reservation request failed", resp);
        if resp is http:ClientRequestError {
            return <http:NotFound> {body: "Unknown hospital, doctor or category"};
        }

        return <http:InternalServerError> {body: resp.message()};
    }
}
```

- Use the `transform` function to transform the payload from a `HealthcareReservation` record to a `HospitalReservation` record.

    ```ballerina
    HospitalReservation hospitalReservation = transform(payload);
    ```

> - Use the transformed payload in the request to send to the hospital service and get the response. Here, The `hospital_id` and `category` values are used as path parameters.

Reviewer: ```suggestion
- Use the transformed payload in the request to the hospital service and get the response. Here, The `hospital_id` and `category` values are used as path parameters.
```

    ```ballerina
    ReservationResponse|http:ClientError resp =
                    hospitalServiceEP->/[payload.hospital_id]/categories/[category]/reserve.post(hospitalReservation);
    ```

- The `log` functions are used to [log](https://ballerina.io/learn/by-example/#log) information at `INFO`, `DEBUG`, and `ERROR` log levels.

> - Use the `is` check to check whether the response is `ReservationResponse` and return it as is (i.e., reservation successful).

Reviewer 2: ```suggestion
- Use the `is` check to check whether the response is a `ReservationResponse` and return it as is (i.e., reservation successful).
```

Reviewer: If we put "a" we have to say "a `ReservationResponse` record"

Author: Yeah. +1 without `a`

Author: fixed via https://github.com/ballerina-guides/integration-tutorials/pull/8/commits/4c858f776cb3ce0a643ec14ad4c0d69b3afa903c

    ```ballerina
    if resp is ReservationResponse {
        log:printDebug("Reservation request successful",
                        name = payload.name,
                        appointmentNumber = resp.appointmentNumber);
        return resp;
    }
    ```

> - If the response is not a `ReservationResponse` record, log the information at `ERROR` level. Return a "NotFound" response if the response from the hospital service is a `http:ClientRequestError` response or an "InternalServerError" response otherwise..

Reviewer 2: ```suggestion
- If the response is not a `ReservationResponse` record, log the information at `ERROR` level. Return a "NotFound" response if the response from the hospital service is a `http:ClientRequestError` response or an "InternalServerError" response otherwise.
```

        "name": "thomas collins",
        "hospital": "grand oak community hospital",
        "category": "surgery",
        "availability": "9.00 a.m - 11.00 a.m",
        "fee": 7000.0
    },
    "patient": {
        "name": "John Doe",
        "dob": "1940-03-19",
        "ssn": "234-23-525",
        "address": "California",
        "phone": "8770586755",
        "email": "johndoe@gmail.com"
    },
    "fee": 7000.0,
    "hospital": "grand oak community hospital",
    "confirmed": false,
    "appointmentDate": "2017-04-02"
}
```

### Concepts covered

- REST API
- HTTP client
- Data mapper

## Develop the application

### Step 1: Set up the workspace

Install [Ballerina Swan Lake](https://ballerina.io/downloads/) and the [Ballerina Swan Lake VS Code extension](https://marketplace.visualstudio.com/items?itemName=wso2.ballerina) on VS Code.

### Step 2: Develop the service

Follow the instructions given in this section to develop the service.

1. Create a new Ballerina project using the `bal` command and open it in VS Code.

```bash
$ bal new transform-message-formats
```

2. Introduce the source code in files with the `.bal` extension (e.g., the `main.bal` file).

Import the

> - `ballerina/http` module to develop the REST API and define the clients that can be used to send requests to the backend services

Reviewer: ```suggestion
- `ballerina/http` module to develop the REST API and define the client that can be used to send requests to the backend services
```

- `ballerina/log` module to log debug, error, or info level information for each client request

```ballerina
import ballerina/http;
import ballerina/log;
```

3. Define two [configurable variables](https://ballerina.io/learn/by-example/#configurability) for the port on which the listener should listen and the URL of the hospital service.

```ballerina
configurable int port = 8290;
configurable string hospitalServiceUrl = "http://localhost:9090";
```

> 4. Define an [`http:Client`](https://ballerina.io/learn/by-example/#http-client) client to send requests to the backend hospital services.

Reviewer: ```suggestion
4. Define an [`http:Client`](https://ballerina.io/learn/by-example/#http-client) client to send requests to the backend hospital service.
```

    ```ballerina
    isolated function transform(HealthcareReservation healthcareReservation) returns HospitalReservation => {
        patient: {
            name: healthcareReservation.name,
            dob: healthcareReservation.dob,
            ssn: healthcareReservation.ssn,
            address: healthcareReservation.address,
            phone: healthcareReservation.phone,
            email: healthcareReservation.email
        },
        doctor: healthcareReservation.doctor,
        hospital: healthcareReservation.hospital,
        appointment_date: healthcareReservation.appointment_date
    };
    ```

6. Define the rest of the records corresponding to the response payload.

```ballerina
type Doctor record {|
    string name;
    string hospital;
    string category;
    string availability;
    float fee;
|};

type ReservationResponse record {|
    int appointmentNumber;
    Doctor doctor;
    Patient patient;
    float fee;
    string hospital;
    boolean confirmed;
    string appointmentDate;
|};
```

- `HealthcareReservation` is the record representing the request payload.
- `HospitalReservation` is the record representing the payload expected by the hospital.

> - `ReservationResponse` is the record representing the payload of the response from the hospital service.

Reviewer: Can remove now?

Reviewer: We can mention the last one in L157 itself IMO.

    isolated resource function post categories/[string category]/reserve(HealthcareReservation payload)
            returns ReservationResponse|http:NotFound|http:InternalServerError {
        HospitalReservation hospitalReservation = transform(payload);

        ReservationResponse|http:ClientError resp =
                    hospitalServiceEP->/[payload.hospital_id]/categories/[category]/reserve.post(hospitalReservation);

        if resp is ReservationResponse {
            log:printDebug("Reservation request successful",
                            name = payload.name,
                            appointmentNumber = resp.appointmentNumber);
            return resp;
        }

        log:printError("Reservation request failed", resp);
        if resp is http:ClientRequestError {
            return <http:NotFound> {body: "Unknown hospital, doctor or category"};
        }

        return <http:InternalServerError> {body: resp.message()};
    }
}
```

- Use the `transform` function to transform the payload from a `HealthcareReservation` record to a `HospitalReservation` record.

    ```ballerina
    HospitalReservation hospitalReservation = transform(payload);
    ```

> - Send the transformed payload to the hospital service and get the response. Here, The `hospital_id` and `category` values are used as path parameters.

Reviewer: ```suggestion
- Use the transformed payload in the request to the hospital service and get the response. Here, The `hospital_id` and `category` values are used as path parameters.
```

    ```ballerina
    ReservationResponse|http:ClientError resp =
                    hospitalServiceEP->/[payload.hospital_id]/categories/[category]/reserve.post(hospitalReservation);
    ```

>     The transformed payload (`hospitalReservation`) is passed directly as an argument to the remote method call.

Reviewer: This sentence is not required IMO.

Author: +1

# Transform message formats

## Overview

> In this tutorial, you will develop a service via which you can reserve appointments at a hospital. The requests are transformed from one format to another and forwarded to the hospital service.

Reviewer: Let's elaborate to say this (the healthcare service) accepts the requests in one format and transforms it to the format expected by the hospital service.

To implement this use case, you will develop a REST service with a single resource using Visual Studio Code with Ballerina Swan Lake extension. The resource will receive the user request, transform the request format into another format, send a request to the hospital service to make a reservation, and respond with the correct reservation details.

The flow is as follows

> 1. Receive a request with a JSON payload similar to the following.

Reviewer: ```suggestion
1. Receive a request with a JSON payload similar to the following.

```

Please fix wherever applicable. Before and after.

```json
{
    "name": "John Doe",
    "dob": "1940-03-19",
    "ssn": "234-23-525",
    "address": "California",
    "phone": "8770586755",
    "email": "johndoe@gmail.com",
    "doctor": "thomas collins",
    "hospital_id": "grandoaks",
    "hospital": "grand oak community hospital",
    "card_no": "7844481124110331",
    "appointment_date": "2017-04-02"
}
```
2. Transform the request data into a form that is similar to the following.

```json
{
    "patient": {
        "name": "John Doe",
        "dob": "1940-03-19",
        "ssn": "234-23-525",
        "address": "California",
        "phone": "8770586755",
        "email": "johndoe@gmail.com",
        "card_no": "7844481124110331"
    },
    "doctor": "thomas collins",
    "hospital": "grand oak community hospital",
    "appointment_date": "2017-04-02"
}
```

3. Extract `hospital_id` field from the JSON payload to construct the URL of the post request to the hospital service.

> 4. Send a request to the hospital service and retrieve the response which will be similar to the following. 

Reviewer: Can be combined IMO.

```json
{
    "appointmentNumber": 8,
    "doctor": {
        "name": "thomas collins",
        "hospital": "grand oak community hospital",
        "category": "surgery",
        "availability": "9.00 a.m - 11.00 a.m",
        "fee": 7000.0
    },
    "patient": {
        "name": "John Doe",
        "dob": "1940-03-19",
        "ssn": "234-23-525",
        "address": "California",
        "phone": "8770586755",
        "email": "johndoe@gmail.com"
    },
    "fee": 7000.0,
    "hospital": "grand oak community hospital",
    "confirmed": false,
    "appointmentDate": "2017-04-02"
}
```

### Concepts covered

- REST API
- HTTP client

> - Data Mapper

Reviewer: ```suggestion
- Data mapper
```
?

```ballerina
type Patient record {|
    string name;
    string dob;
    string ssn;
    string address;
    string phone;
    string email;
|};

type RequestData record {|
    *Patient;
    string doctor;
    string hospital_id;
    string hospital;
    string card_no;
    string appointment_date;
|};

type ReservationRequest record {|
    Patient patient;
    string doctor;
    string hospital;
    string appointment_date;
|};

type Doctor record {|
    string name;
    string hospital;
    string category;
    string availability;
    float fee;
|};

type ReservationResponse record {|
    int appointmentNumber;
    Doctor doctor;
    Patient patient;
    float fee;
    string hospital;
    boolean confirmed;
    string appointmentDate;
|};
```

- `RequestData` is the type of the data the resource takes as the payload.
- `ReservationRequest` is the type of the data that will be sent to the hospital service.
- `ReservationResponse` is the type of the response received from the hospital service.

6. Use data-mapper to define the function which transform the payload from `RequestData` to `ReservationRequest`. 

>     Follow the guide shown below to create the `transform` function.

Reviewer: ```suggestion
6. Use the data mapper to define the `transform` function which transforms the payload from `RequestData` to `ReservationRequest` as follows. 
```

>     ![Data Mapper Guide](./resources/tutorial_transform_message_formats_data_mapper_guide.gif)

Reviewer: ```suggestion
    ![Data mapper guide](./resources/tutorial_transform_message_formats_data_mapper_guide.gif)
```

Fix wherever applicable.

>     The data mapper diagram of the `transform` function is shown below.

Reviewer: ```suggestion
    The data mapper view of the completed `transform` function is shown below.
```

    ![Data Mapper Diagram](./resources/tutorial_transform_message_formats_data_mapper_diagram_transform_function.png)

>     The `transform` function will finally look like this:

Reviewer: ```suggestion
    The source code of the `transform` function will be as follows.
```

    ```ballerina
    isolated function transform(RequestData details) returns ReservationRequest => {
        patient: {
            name: details.name,
            dob: details.dob,
            ssn: details.ssn,
            address: details.address,
            phone: details.phone,
            email: details.email
        },
        doctor: details.doctor,
        hospital: details.hospital,
        appointment_date: details.appointment_date
    };
    ```

7. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resource that accepts user requests, makes a call to the backend hospital service, retrieves reservation details, and responds to the client.

- Use `/healthcare` as the service path (or the context) of the service which is attached to the listener listening on port `port`. 

- The HTTP resource allows the `POST` operation on resource path `/categories/{category}/reserve`, where `category` (corresponding to the specialization) is a path parameter.

- Use `RequestData` as a parameter indicating that the resource expects a JSON object corresponding to `RequestData` as the payload. 

> - Use `ReservationResponse|http:NotFound|http:BadRequest|http:InternalServerError` as the return type to indicate that the response will have a JSON payload corresponding to `ReservationResponse` on success or the response will be a "NotFound", "BadRequest" or "InternalServerError" response on error.

Reviewer: Move to after the snippet.

    ```ballerina
    ReservationResponse|http:ClientError resp = hospitalServiceEP->/[hospitalId]/categories/[category]/reserve.post(req);
    ```

- The `log` functions are used to [log](https://ballerina.io/learn/by-example/#log) information at `INFO`, `DEBUG`, and `ERROR` log levels.

- Use the `is` check to check whether the response is `ReservationResponse` and return it as is (i.e., reservation successful).

    ```ballerina
    if resp is ReservationResponse {
        log:printDebug("Reservation request successful",
                        name = payload.name,
                        appointmentNumber = resp.appointmentNumber);
        return resp;
    }

- If the response is not a `ReservationResponse`, log the information at `ERROR` level. Return a "NotFound" response if the response is a `http:ClientRequestError` or an "InternalServerError" response if otherwise.

    ```ballerina
    log:printError("Reservation request failed", resp);
    if resp is http:ClientRequestError {
        return <http:NotFound> {body: "Unknown hospital, doctor or category"};
    }

    return <http:InternalServerError> {body: resp.message()};
    ```

#### Complete source code

> You have successfully developed the required service.

Reviewer: Still in the wrong place?

# Transform message formats

## Overview

> In this tutorial, you will develop a service via which you can reserve appointments at hospitals. The requests are transformed from one format to another and forwarded to the hospital service.

Reviewer: ```suggestion
In this tutorial, you will develop a service via which you can reserve appointments at hospitals. The requests are transformed from one format to another and forwarded to the hospital service to reserve the appointment.
```

> To implement this use case, you will develop a REST service with a single resource using Visual Studio Code with Ballerina Swan Lake extension. The resource will receive the user request, transform the request format into a format expected by the hospital service, send a request to the hospital service to make a reservation, and respond with the correct reservation details.

Reviewer: ```suggestion
To implement this use case, you will develop a REST service with a single resource using Visual Studio Code with the Ballerina Swan Lake extension. The resource will receive the user request, transform the payload to the format expected by the hospital service, send a request with the transformed payload to the hospital service to make a reservation, and respond with the reservation details.
```

The flow is as follows

1. Receive a request with a JSON payload similar to the following.

```json
{
    "name": "John Doe",
    "dob": "1940-03-19",
    "ssn": "234-23-525",
    "address": "California",
    "phone": "8770586755",
    "email": "johndoe@gmail.com",
    "doctor": "thomas collins",
    "hospital_id": "grandoak",
    "hospital": "grand oak community hospital",
    "card_no": "7844481124110331",
    "appointment_date": "2017-04-02"
}
```

> 2. Transform the request data into a form that is similar to the following.

Reviewer: ```suggestion
2. Transform the request payload to the following form.
```

```bash
$ bal new transform-message-formats
```

2. Introduce the source code in files with the `.bal` extension (e.g., the `main.bal` file).

Import the
- `ballerina/http` module to develop the REST API and define the clients that can be used to send requests to the backend services
- `ballerina/log` module to log debug, error, or info level information for each client request

```ballerina
import ballerina/http;
import ballerina/log;
```

3. Define two [configurable variables](https://ballerina.io/learn/by-example/#configurability) for the port on which the listener should listen and the URL of the hospital service.

```ballerina
configurable int port = 8290;
configurable string hospitalServiceUrl = "http://localhost:9090";
```

4. Define an [`http:Client`](https://ballerina.io/learn/by-example/#http-client) client to send requests to the backend hospital services.

```ballerina
final http:Client hospitalServiceEP = check initializeHttpClient();

function initializeHttpClient() returns http:Client|error => new (hospitalServiceUrl);
```

> **Note** The argument passed to the `new` expression (`hospitalServiceUrl`) is the URL of the backend hospital service.
> 

> > Here, a separate function is used to initialize the clients to aid with testing. Alternatively, the `new` expression can be used directly to initialize the clients.

Reviewer: ```suggestion
> Here, a separate function is used to initialize the client to aid with testing. Alternatively, the `new` expression can be used directly to initialize the client.
```

>
> ```ballerina
> final http:Client hospitalServiceEP = new (hospitalServiceUrl);
>```

> 5. Use the data-mapper to define the `transform` function which transform the payload from `HealthcareReservation` to `HospitalReservation`. 

Reviewer: ```suggestion
5. Use the data mapper to define the `transform` function which transforms a `HealthcareReservation` record, representing the payload, to a `HospitalReservation` record. 
```

>     Follow the guide shown below to create the `transform` function.

Reviewer: This is not required IMO.

Reviewer: And not necessarily a guide.

    ![Data mapper guide](./resources/tutorial_transform_message_formats_data_mapper_guide.gif)

    The data mapper view of the completed `transform` function is shown below.

    ![Data mapper view](./resources/tutorial_transform_message_formats_data_mapper_view.png)

    The source code of the `transform` function will be as follows.

    ```ballerina
    isolated function transform(HealthcareReservation healthcareReservation) returns HospitalReservation => {
    patient: {
        name: healthcareReservation.name,
        dob: healthcareReservation.dob,
        ssn: healthcareReservation.ssn,
        address: healthcareReservation.address,
        phone: healthcareReservation.phone,
        email: healthcareReservation.email
    },
    doctor: healthcareReservation.doctor,
    hospital: healthcareReservation.hospital,
    appointment_date: healthcareReservation.appointment_date

> };

Reviewer: Indentation is off.

    ```

> 6. Define rest of the records corresponding to the response payloads. The final set of records will look like this.

Reviewer: ```suggestion
6. Define the rest of the records corresponding to the response payload.
```

> 

Reviewer: I would not include the rest of the records (already generated ones) here. Since they will be available in the complete source section anyway.

```ballerina
type HealthcareReservation record {|
    string name;
    string dob;
    string ssn;
    string address;
    string phone;
    string email;
    string doctor;
    string hospital_id;
    string hospital;
    string card_no;
    string appointment_date;
|};

type Patient record {|
    string name;
    string dob;
    string ssn;
    string address;
    string phone;
    string email;
|};

type HospitalReservation record {|
    Patient patient;
    string doctor;
    string hospital;
    string appointment_date;
|};

type Doctor record {|
    string name;
    string hospital;
    string category;
    string availability;
    float fee;
|};

type ReservationResponse record {|
    int appointmentNumber;
    Doctor doctor;
    Patient patient;
    float fee;
    string hospital;
    boolean confirmed;
    string appointmentDate;
|};
```

> - `HealthcareReservation` is the type of the payload the resource takes as the payload.

Reviewer: Incorrect sentence.

```suggestion
- `HealthcareReservation` is the record representing the request payload
```

> - `HospitalReservation` is the type of the payload that the hospital service expects.

Reviewer: ```suggestion
- `HospitalReservation` is the record representing the payload expected by the hospital service
```

> - `ReservationResponse` is the type of the response that the hospital service returns.

Reviewer: ```suggestion
- `ReservationResponse` is the record representing the payload of the response from the hospital service
```

> 7. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resource that accepts user requests, makes a call to the backend hospital service to retrieve reservation details, and responds to the client.

Reviewer: ```suggestion
7. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resource that accepts user requests, transforms the payload, makes a call to the backend hospital service to make the reservation, and responds to the client.
```

```ballerina
service /healthcare on new http:Listener(port) {
    isolated resource function post categories/[string category]/reserve(HealthcareReservation payload)
            returns ReservationResponse|http:NotFound|http:BadRequest|http:InternalServerError {
        
   }
}
```

- Use `/healthcare` as the service path (or the context) of the service which is attached to the listener listening on port `port`. 

- The HTTP resource allows the `POST` operation on resource path `/categories/{category}/reserve`, where `category` (corresponding to the specialization) is a path parameter.

- Use `HealthcareReservation` as a parameter indicating that the resource expects a JSON object corresponding to `HealthcareReservation` as the payload. 

> - Use `ReservationResponse|http:NotFound|http:BadRequest|http:InternalServerError` as the return type to indicate that the response will have a JSON payload corresponding to `ReservationResponse` on success or the response will be a "NotFound", "BadRequest" or "InternalServerError" response on error.

Reviewer: ```suggestion
- Use `ReservationResponse|http:NotFound|http:BadRequest|http:InternalServerError` as the return type to indicate that the response will have a JSON payload corresponding to `ReservationResponse` on success or the response will be a "NotFound", "BadRequest", or "InternalServerError" response on error.
```

Won't be required when the unused one is removed though.

8. Implement the logic.

```ballerina
service /healthcare on new http:Listener(port) {
    isolated resource function post categories/[string category]/reserve(HealthcareReservation payload)
            returns ReservationResponse|http:NotFound|http:BadRequest|http:InternalServerError {
        HospitalReservation reservationReq = transform(payload);

        ReservationResponse|http:ClientError resp =
                    hospitalServiceEP->/[payload.hospital_id]/categories/[category]/reserve.post(reservationReq);

        if resp is ReservationResponse {
            log:printDebug("Reservation request successful",
                            name = payload.name,
                            appointmentNumber = resp.appointmentNumber);
            return resp;
        }

        log:printError("Reservation request failed", resp);
        if resp is http:ClientRequestError {
            return <http:NotFound> {body: "Unknown hospital, doctor or category"};
        }

        return <http:InternalServerError> {body: resp.message()};
    }
}
```

> - Use the `transform` function to transform the data from `HealthcareReservation` to `HospitalReservation`.

Reviewer: ```suggestion
- Use the `transform` function to transform the payload from a `HealthcareReservation` record to a `HospitalReservation` record.
```

    ```ballerina
    HospitalReservation reservationReq = transform(payload);
    ```

- Send the transformed payload to the hospital service and get the response. Here, The `hospital_id` and `category` values are used as path parameters.

    ```ballerina
    ReservationResponse|http:ClientError resp =
                    hospitalServiceEP->/[payload.hospital_id]/categories/[category]/reserve.post(reservationReq);
    ```

>     The payload is specified directly as an argument to the remote method call.

Reviewer: Incorrect?

- The `log` functions are used to [log](https://ballerina.io/learn/by-example/#log) information at `INFO`, `DEBUG`, and `ERROR` log levels.

- Use the `is` check to check whether the response is `ReservationResponse` and return it as is (i.e., reservation successful).

    ```ballerina
    if resp is ReservationResponse {
        log:printDebug("Reservation request successful",
                        name = payload.name,
                        appointmentNumber = resp.appointmentNumber);
        return resp;
    }
    ```

> - If the response is not a `ReservationResponse`, log the information at `ERROR` level. Return a "NotFound" response if the response is a `http:ClientRequestError` or an "InternalServerError" response if otherwise.

Reviewer: ```suggestion
- If the response is not a `ReservationResponse` record, log the information at `ERROR` level. Return a "NotFound" response if the response from the hospital service is a `http:ClientRequestError` response or an "InternalServerError" response otherwise.
```

            return resp;
        }

        log:printError("Reservation request failed", resp);
        if resp is http:ClientRequestError {
            return <http:NotFound> {body: "Unknown hospital, doctor or category"};
        }

        return <http:InternalServerError> {body: resp.message()};
    }
}

isolated function transform(HealthcareReservation healthcareReservation) returns healthcareReservation => {
    patient: {
        name: healthcareReservation.name,
        dob: healthcareReservation.dob,
        ssn: healthcareReservation.ssn,
        address: healthcareReservation.address,
        phone: healthcareReservation.phone,
        email: healthcareReservation.email
    },
    doctor: healthcareReservation.doctor,
    hospital: healthcareReservation.hospital,
    appointment_date: healthcareReservation.appointment_date
};
```

#### Diagram

The [sequence diagram view](https://wso2.com/ballerina/vscode/docs/implement-the-code/sequence-diagram-view/) for the implemented resource method is the following.

![sequence diagram](./resources/tutorial_transform_message_formats_diagram.png)

### Step 3: Build and run the service

> You can run this service by using `bal run` command after navigating to the project root.

Reviewer: We decided to keep navigating first, right?

```suggestion
You can run this service by navigating to the project root and using the `bal run` command.
```

```bash
transform-message-formats$ bal run
Compiling source

>         integration_tutorials/transform-message-formats:0.1.0

Reviewer: Incorrect? Underscores?
